### PR TITLE
Assignment 5 - Hunter Thompson & Hunter Allen

### PR DESCRIPTION
**Pull request by StrangeRanger**
@hallen10

### CODE REVIEW

**Comment by Malcolm-Shepherd** [comment]
## Instructions

- Create a *class library* project called "GenericsHomework.". ✔
- Create a node class that can contain a value of any type and points to the next node and traversing the next node points back to the first item.
  - Define the `Node` class✔
  - The values should be homogeneous.✔
  - Include a constructor that takes a value.  (No validation is necessary on the value). ✔
  - Add a `ToString()` override that writes out the value's `ToString()` result. ❌
  - Add a `Next` property that references the next node or else refers back to itself if there are no other nodes in the list. ✔
    - The `Next` property should be non-nullable (careful to follow the non-nullable property guidelines) ❌
    - The `Next` property setter should be private. ✔
  - Add an `Append` method that takes a value and appends a new `Node` instance after the current node (by invoking the `Next` property). ❌
  - Add a Clear method that effectively removes all items from a list except the current node. Pay attention as to whether you should be concerned with the following:
    - Whether it is sufficient to only set Next to itself ✔
    - Whether to set the removed items to circle back on themselves. In other words, whether to close the loop of the removed items. (Provide a test to show why this is required if it is required). ✔
    - Given there is a circular list of items, provide a comment to indicate whether you need to worry about garbage collection because all the items point to each other and therefore may never be garbage collected. ✔
  - Create an Exists method to test to see if a value exists in the list. ✔
  - Throw an error on an attempt to Append a duplicate value. (Make sure you test for this case) ✔
- You should not rely on any BCL generic classes for your implementation. ✔
## Fundamentals

- Place all shared project properties into a `Directory.Build.Props` file.✔
- Place all shared project items into a `Directory.Build.targets` file.✔
- nullable reference types is enabled  ✔
- Ensure that you turn on code analysis for all projects(EnableNETAnalyzers)  ✔
- Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional)   ✔
- and enabled .NET analyzers for both projects ✔
- For this assignment, always use `Assert.AreEqual<T>()` (the generic version)  ✔
- All of the above should be unit tested ✔
- Choose simplicity over complexity ✔

File: GenericsHomework/Node.cs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    public void Append(T value)
+    {
+        if (Exists(value))
+        {
+            throw new DuplicateDataInArrayException(nameof(value));
+        }
+
+        Node<T> current = this;
+
+        while (current.Next != current) { current = current.Next; }
```
Comment (Malcolm-Shepherd) [inline GenericsHomework/Node.cs:24]: This Appends the new Node at the end of the list, not after the current node

File: GenericsHomework/Node.cs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+            count++;
+        }
+
+        return count;
+    }
+    */
+
+    public override string ToString()
+    {
+        string holder = "{ ";
```
Comment (JessicaH46) [inline GenericsHomework/Node.cs:72]: For your ToString, I would simplify the code to something similar to this
`if(Value != null) return Value.ToString();
            else return string.Empty;`
this is what we had for ours and it did the job!

**Comment by JessicaH46** [comment]
Create a class library project called "GenericsHomework.". ✔
Create a node class that can contain a value of any type and points to the next node and traversing the next node points back to the first item.
Define the Node class
The values should be homogeneous.
Include a constructor that takes a value. (No validation is necessary on the value). ✔
Add a ToString() override that writes out the value's ToString() result. ❌✔ * Could simplify, left a comment
Add a Next property that references the next node or else refers back to itself if there are no other nodes in the list. ✔
The Next property should be non-nullable (careful to follow the non-nullable property guidelines) ❌
The Next property setter should be private. ✔
Add an Append method that takes a value and appends a new Node instance after the current node (by invoking the Next property). ❌✔
Add a Clear method that effectively removes all items from a list except the current node. Pay attention as to whether you should be concerned with the following:
Whether it is sufficient to only set Next to itself ✔
Whether to set the removed items to circle back on themselves. In other words, whether to close the loop of the removed items. (Provide a test to show why this is required if it is required). ✔
Given there is a circular list of items, provide a comment to indicate whether you need to worry about garbage collection because all the items point to each other and therefore may never be garbage collected. ✔
Create an Exists method to test to see if a value exists in the list. ✔
Throw an error on an attempt to Append a duplicate value. (Make sure you test for this case) ✔
You should not rely on any BCL generic classes for your implementation. ✔

Overall you have all the main components! The fundamentals all look up to date and your tests all look good to me! Great job! I did leave a comment on your toString with a suggestion, I also like how you checked to see if the data was laready in the list with your append! I didn't think to do it that way, I saw someone else's comment on your append, but ours looked pretty similar except for at the end of your loop we had an extra line that would be 
current.Next.Next = this;
Not sure if that's what would resolve your problem, but it might so I will leave it here. Good job on the assignment!

File: GenericsHomework/Node.cs (hunk 1 above)
Comment (StrangeRanger) [reply GenericsHomework/Node.cs:24]: Is that not the point? If I do it after the current node, it'd be constantly overriding the second node in the list.

File: GenericsHomework/Node.cs [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        Node<T> current = this;
+
+        while (current.Next != current)
+        {
+            holder += current.Value + ", ";
+            current = current.Next;
+        }
+
+        holder += current.Value + " }";
+        return holder;
```
[the remark this answers is not in the document]
Comment (StrangeRanger) [reply GenericsHomework/Node.cs]: Thank you for your suggestion here.

File: GenericsHomework/Node.cs [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+public class Node<T>
+{
+    // Must be non-nullable.
+    public Node(T value)
+    {
+        Value = value !;
+        Next = this;
+    }
+
+    public Node<T> Next { get; private set; }
```
[the remark this answers is not in the document]
Comment (StrangeRanger) [reply GenericsHomework/Node.cs:11]: Thanks for pointing that out

File: GenericsHomework/Node.cs (hunk 2 above)
Comment (StrangeRanger) [reply GenericsHomework/Node.cs]: That's a good idea to have it check if the value is null. I implemented it as such: `return Value is not null ? Value.ToString()! : string.Empty;`

**Review by mmwoodfo — COMMENTED** [review]
Good job!

**Pull request closed without merging**